### PR TITLE
Unity7-esk backlight items support

### PR DESCRIPTION
**Pull request by SavageTiger**
Hi,

Since Unity7 is sort-of discontinued I switched to Gnome3, dash to dock provides most of the missing 
functionality for me.

However I always liked the look of the colorful side-bar that clearly indicated what applications are active. So I tried to replicate this functionality in dash-to-dock and I  think I succeeded.

example

I added a toggle to the Appearance tab to enable / disable the functionality.

I hope you like it.

* There is a small issue where you can only go up to 48px max icon size, since there seems to be a SVG scaling issue in Gtk itself, but other than that it works flawlessly.

### CODE REVIEW

**Comment by micheleg** [comment]
Hi Sven,

thanks a lot for your contribution. I quickly tested it your code and it seems to run smoothly. It's a neat effect which adds quite a lot of colour to the shell! Not sure I personally prefer it over a more simple default theme, but I understand the appealing for people coming from Ubuntu.

Regarding the  styling I have always being a bit careful to not trying to implement every possible customization, preferring instead to leave as much as possible to the theme writers. Not everything can however be controlled by css, and you matching colouring of  the icons background is a good example.

I am considering enabling additional styles for the running/focused applications (See #434), in this would reasonably fit in that direction. I'll try to review with a bit of care your code (it seems reasonable at a first glance).

File: utils.js [hunk 1]
```diff
@@ -83,6 +83,81 @@ const GlobalSignalsHandler = new Lang.Class({
...
+
+        return {
+            h: h,
+            s: s,
+            v: v
+        };
+    }
+});
+
+/**
```
Comment (micheleg) [inline utils.js]: I'm just wondering if Clutter color functions could be used instead?

https://developer.gnome.org/clutter/1.26/clutter-Colors.html#clutter-color-to-hls
https://developer.gnome.org/clutter/1.26/clutter-Colors.html#clutter-color-from-hls

Comment (SavageTiger) [inline utils.js]: Hi Michele,
Thank you for your quick and enthusiastic response! I am happy that there is interest in my changes, I will investigate if I can use the HLS native code (since its not the same as HSV)

Comment (micheleg) [inline utils.js]: It's not a problem to have those functions,  unless there are simple equivalent already available.

**Comment by micheleg** [comment]
I've noticed some performance issue with your branch, in particular when changing workspace, with the shell hanging for ~1s. The same occurs at the first activation of the option from the settings.

**Comment by SavageTiger** [comment]
I think that has to do with having to calculate the color of all items every time they are created, maybe i can fix that by adding named caching. What hardware are you using? I am testing on a pretty new system (ryzen 7) so I dont really notice any slowdown.

**Comment by micheleg** [comment]
I'm on a quite new laptop (i5/skylake). Now, restarting the shell I noticed the the slowdown is gone, but it was very evident before and clearly dependent on the settings for this feature being on or off. I'll see if the problem reappears.

**Comment by SavageTiger** [comment]
Update; I have managed to replicate the laggy behaviour on my local machine, i'll investigate further, see if i can fix it.

**Comment by micheleg** [comment]
I noticed that the glossy svg is missing in the makefile.

**Comment by micheleg** [comment]
I found out that the slow down occurs with specific applications. I noticed it with slack. I noticed that it use a large icon (512x512px), even when used at lower scale. The issue could therefore be the loop over the image pixels? The colour could maybe be cached, but probably its calculation should be optimized. the simplest thing that pass through my mind is sampling only a subset (random/grid?) of the pixels.

**Comment by micheleg** [comment]
Also,

the glossy background works correctly only for icons size of 48 or less. Here for instance the result at icon size 64:
screenshot from 2017-06-02 22-59-26

**Comment by SavageTiger** [comment]
Maybe the glossy background should be removed, as I said in my original pull request, if you scale your images higher then 48pixels, Gtk stops properly scaling the svg's I did find some old bug reports that do not clearly seem to indicate if the problem has been fixed or even related. What do you think?

https://bugzilla.gnome.org/show_bug.cgi?id=316825
https://bugzilla.gnome.org/show_bug.cgi?id=590663

Also i ran into a different issue that my electron apps (jira and whatsapp) seem to always use a fallback, the problem here is that in those (and most likely also other cases) the shell_app_create_icon_texture() in https://github.com/GNOME/gnome-shell/blob/master/src/shell-app.c also has a fallback method that will return a ClutterTexture. I have been trying for hours to read pixels from a ClutterTexture but so far I failed, I did manage to pass it to a clutter stage and call read_pixels on that, but that made Gtk crash doe to this bug: https://bugzilla.gnome.org/show_bug.cgi?id=653590 My electron apps are made using natifier so I am not sure of Slack also has this problem since I am not a slack user. Maybe someone else knows a solution for this, since I am at the point going crazy, but since it rarely happens maybe its not even worth the affort, since there is a sane fallback option, that could be improved by adding a fallback color picker to the settings in the future.

As for the performance problem, I have tested using a local cache and had not had problems as of yet, I am at the office tomorrow, and will publish the updated source code then.

**Comment by micheleg** [comment]
Those bugs are for gnome-panel(gtk), while the shell uses Clutter/St so I'm not sure they aply here.

Adding this css rule seems to fix the background scaling:
 ```background-size: contain;```

Should we keep it or not. I'm not sure. It's very unity like, which I guess was your goal. Without the glossy background, the contrast between the running apps with the colourful background and the non running apps seems a bit ugly to me.

**Comment by SavageTiger** [comment]
Added caching, and fixed the svg resizing issue

File: appIcons.js [hunk 2]
```diff
@@ -297,6 +312,13 @@ const MyAppIcon = new Lang.Class({
             this.parent();
         this._onFocusAppChanged();
         this._updateCounterClass();
+        
+        // Enable / Disable the backlight of running apps
+        if (this.app.state !== Shell.AppState.STOPPED && this._dtdSettings.get_boolean('unity-backlit-items') === true) {
```
Comment (franglais125) [inline appIcons.js]: Could we also check for the number of windows? This would make the feature compatible with workspace isolation!

`&& getInterestingWindows(this.app, this._dtdSettings).length > 0`

**Comment by SavageTiger** [comment]
Done and done :D

File: appIcons.js [hunk 3]
```diff
@@ -689,11 +695,14 @@ const MyAppIcon = new Lang.Class({
 
         let rgb = colorUtils.HSVtoRGB(hsv.h, hsv.s, hsv.v);
 
-        return {
+        // Cache the result.
+        this._backgroundColors = {
```
Comment (franglais125) [inline appIcons.js]: I was trying to figure out why the caching was not working... It turns out the variable here ends with an "s". So the color is actually never cached!

```
this._backgroundColor = null;
this._backgroundColors = {
```

**Comment by franglais125** [comment]
I don't know how reasonable this is, but here [659982c012e7acffe36bfb39653c0805341d03ef] is a global approach to the caching mechanics. I've made it global, so even if an application is closed (or if switching workspaces with isolation enabled) the cache is kept.

For applications like Slack, the shell will only freeze the first time it's opened. A sub-sampling would probably still be needed for weak cpus.

**Comment by franglais125** [comment]
Here is a patch for resampling [71e438ade18bcef2679a17b21d0faf4ba842449a].
Below you can find screenshots, where the only affected icon is Slack, as the resampling is done only for 256 and 512px.

No resampling:
screenshot from 2017-06-10 16-04-54

Moderate resampling (**patch above**):
screenshot from 2017-06-10 16-04-54

Aggressive resampling (resampling twice as hard, this seems too much):
screenshot from 2017-06-10 16-07-29


I'm getting used to this, and I am sort of liking it. Awesome work @SavageTiger !

**Comment by SavageTiger** [comment]
Wow nice work fixing that performance issue!, I merged it into my branch, and did some readability refactoring. My dots still have the cool two tone effect that is missing from your screenshots, do you know why this is?

File: appIcons.js [hunk 4]
```diff
@@ -747,6 +741,32 @@ const MyAppIcon = new Lang.Class({
...
+     * improve performance.
+     *
+     * @param pixBuf
+     * @param pixels
+     * @param resampleX
+     * @param resampleY
+     *
+     * @return [];
+     */
+    _resamplePixels: function (pixels, pixBuf, resampleX, resampleY) {
```
Comment (franglais125) [inline appIcons.js]: Thanks for the refactoring!

Small bug: the `pixels` and `piBuf` inputs are swapped.
```
_resamplePixels: function (pixels, pixBuf, resampleX, resampleY)
_resamplePixels: function (pixBuf, pixels, resampleX, resampleY)
```

Comment (SavageTiger) [inline appIcons.js]: Ow oops, good catch, I dont have any large icons so it was hard for me to test :D

**Comment by franglais125** [comment]
@micheleg @SavageTiger I tried to profile the `_calculateColorPalette` function, to see how much the resampling is helping.

Using the Slack icon (only one I could find with 512px):
- 512 px: takes 3665ms
- 256 px: takes 938ms
- 128 px: takes 240ms
- 64 px: takes 60ms

The result is very linear, as expected. The resampling included in the code now downsamples 512px and 256px icons to 64px.

I got these results on a Sandybridge dual core laptop (~2011), so not exactly something new nor too powerful. At 512, it took over 3s to calculate the icon. So yes, that was too much. The resampling plus the global cache should help a lot in this respect.

Here is the code I used (line 598):
```
        global.log('Calculating colorPalette for: ' + this.app.get_id());
        let now = new Date();
        let ms = now.getTime();
        let colorPallete = this._calculateColorPalette(pixBuf);
        now = new Date();
        ms = now.getTime() - ms;
        global.log('Done in ' + ms+ 'ms');
```

**Comment by SavageTiger** [comment]
My results using the Slack icon, On a Ryzen7 1700x (clocked at 3.9)

- Without re-sampling: 1249ms
- With re-sampling: 37ms
- With cache (after the first call draw-call off course): 0ms

File: appIcons.js [hunk 5]
```diff
@@ -542,6 +574,199 @@ const MyAppIcon = new Lang.Class({
...
+            return null;
+        }
+
+        if (iconTexture instanceof Gio.FileIcon) {
+            // Use GdkPixBuf to load the pixel buffer from the provided file path
+            return GdkPixbuf.Pixbuf.new_from_file(iconTexture.get_file().get_path());
+        }
+        
+        // Get the pixel buffer from the icon theme
+        return themeLoader.load_icon(iconTexture.get_names()[0], 24, 0);
```
Comment (franglais125) [inline appIcons.js]: Small suggestion regarding the default size loaded. If we increase it from 24 to 64, there is quite an impact in the end result. This is of course just a suggestion! Cheers.

24px:
Small suggestion regarding the default size requested. If we increase it from 24 to 64, there is quite an impact in the end result.

24px:
screenshot from 2017-06-12 12-19-00

48px:
screenshot from 2017-06-12 12-19-40

64px:
screenshot from 2017-06-12 12-20-05

Comment (SavageTiger) [inline appIcons.js]: I agree, after adding cache I think that would have minimal impact, and gives us better results.

**Comment by franglais125** [comment]
To help with the (eventual I hope) merge of this branch, I performed a rebase against Michele's master.
https://github.com/franglais125/dash-to-dock/commits/unity7_rebased

I certainly don't want to takeover this PR. Just trying to get it to a merge-able state!

Commits dropped:
- [31489d9] Messy prototype of unity-esk functionality and [e6818ff] Revert "Messy prototype of unity-esk functionality" , as they both cancel out.
- [84d47d6] Simplify Settings.ui changeset. This was squashed so that the changeset on Settings.ui is overall small.
- [b00c6ea] Merge branch 'master' into master. This is no longer necessary.
- [d3970c0] Merged with upstream, fixed mergeconflicts. This is no longer necessary

**Comment by franglais125** [comment]
When resampling, half the time of the whole calculation is actually doing the resampling.

I managed to speed up the resampling by a factor of ~5. On my slow laptop, I can now render the (resampled) 64px in ~35ms, down from 60ms.

Here is the commit: [a282660].

**Comment by franglais125** [comment]
And finally, a new optimization, this time on the main loop for the color calculation. It is now 3 times faster! [ae7d3fa91f3a6024a5ff9d8c43e0b905937fcc94].

64px icon background calculated in 11ms (14ms if resampled). Very reasonable IMO.

As a reminder, this is only the first time the color is computed, then it takes 0ms as it's cached.

**Comment by SavageTiger** [comment]
Wow good work man, I did not realize js recalculates the upperbound every iteration, makes sense I suppose.
And thank you for all the help making this better. Dont worry I understand this is not a hostile takeover.  I'd say its more like teamwork :+1:

**Comment by SavageTiger** [comment]
I just forcepushed your rebased branch into my forked master (I just realized without context this is a weird sentence). Hope the PR is still valid

**Comment by micheleg** [comment]
Nice work!

One note: at the moment this option  doesn't get disabled when "Use built-in theme" is enabled.

**Comment by micheleg** [comment]
I'm thinking of restructuring the code handling the application status style, in particular with the prospect of enabling multiples style options, including this and possible importing some from dash-to-panel (#434). To do this I tried to separate the logic to a dedicated class. I've started working on it on this branch:

https://github.com/micheleg/dash-to-dock/tree/appIconIndicators.

At the moment I'm still playing around to see if it actually makes sense, to be sure to avoid over-engineering the thing, so don't rebase on it.  I just wanted to let you know. Happy to get suggestions.

**Comment by franglais125** [comment]
@micheleg That looks very promising. Just like the docking.js case when introducing multi-monitor, this is a welcome structuring of the code!

**Comment by SavageTiger** [comment]
@micheleg I dont hhink that is a bad plan, stricly speaking they are there own entity.
While i was working on this I was experimenting with drawing the indicators in different shapes (arrow in this case, same Unity), but after changing the color and slightly moving them I decided that it was Ok for now.

**Comment by franglais125** [comment]
I pushed two commits to https://github.com/franglais125/dash-to-dock/commits/unity7_rebased:

- https://github.com/franglais125/dash-to-dock/commit/9ace4325aad24c722d8caee69aa3e055a60de1ae: small optimization
- https://github.com/franglais125/dash-to-dock/commit/3ea0a6d3df670e5108787a58b2e23344cff21cb0 fixes the incompatibility with "Use built-in theme".

File: appIcons.js [hunk 6]
```diff
@@ -623,6 +783,12 @@ const MyAppIcon = new Lang.Class({
         let spacing = radius + borderWidth; // separation between the dots
         let n = this._nWindows;
 
+        // Lightly adjust the styling in case of unity7 backlit mode
+        if (this._dtdSettings.get_boolean('unity-backlit-items') === true) {
```
Comment (franglais125) [inline appIcons.js]: Maybe here we can check if `custom-theme-customize-running-dots` is true, and not override the padding and width from the settings. So something like:
```
if (this._dtdSettings.get_boolean('unity-backlit-items') &&
    !this._dtdSettings.get_boolean('custom-theme-customize-running-dots')) {
        padding = 1.45;
        borderWidth = 2;
}
```

**Comment by nsrosenqvist** [comment]
Sorry for just jumping in here, but is it possible to have the glossiness as a separate option? I'd love to have the backlight feature but as a solid color and square, looks more modern. Or is it customizable with CSS?

**Comment by franglais125** [comment]
Here is how it looks without the glossy effect:
screenshot from 2017-06-20 18-17-45


Maybe we could have a drop-down menu, with the options `[off, glossy, simple]` ?

**Comment by SavageTiger** [comment]
@franglais125 Sorry it took so long for me to get back to you, there is allot going on at the moment. I just merged your changes, thanks for that.

**Comment by franglais125** [comment]
@SavageTiger no worries, I'm just pushing ideas/suggestions. You certainly don't need to pull everything from that branch :)

**Comment by tliron** [comment]
Many thanks to @SavageTiger for this implementation! I have borrowed your code (with credit, of course) for the extension I am working on: https://github.com/tliron/em-dash

**Comment by SavageTiger** [comment]
@tliron You are very welcome. If you ever find out how to get the pixBuf if ClutterTextures (currently this PR uses a fallback), please let me know

**Comment by micheleg** [comment]
@tliron very nice project, I'll definitely dig into it for inspiration!

**Comment by tliron** [comment]
@micheleg It's the opposite, I was mostly inspired by you. :) If you or @SavageTiger would like to be credited with your real names, please let me know. For now I'm just using the names I have.

I will look into that problem, @SavageTiger. It's indeed not so easy because Clutter abstracts that away. With `GIcon`, too, there are a few different types that are available, but really because of GObject anyone can create their own implementation. (For example, someone might create a GIcon that renders dynamically!). By the way, you support `FileIcon`, but I don't you are checking specifically for `ThemedIcon`. Also, there is `BytesIcon`. I never saw one used, but who knows?

**Comment by tliron** [comment]
Unfortunately, it seems impossible to access the ClutterTexture data directly, at least in JavaScript. Internally it uses Cogl for everything, but CoglHandle is not exposed via GObject introspection. It would be possible to write something in C to do this, but that's obviously not in our scope.

**Comment by tliron** [comment]
OK! So I found a better way by looking in the `St.TextureCache` source code. You don't want to use `Clutter.Texture` at all, because actually St hands over the pixbuf data to it. Instead you want to use the `Gtk.IconTheme` to load it directly, essentially duplicating what St does. See my [solution](https://github.com/tliron/em-dash/blob/master/utils/icon.js).

In a perfect world, it might seem nice if `St.TextureCache` would expose what it already loaded so that we can reuse it. But actually, it might not be best for us. For example, if the dash shows very tiny icons, we won't have enough pixels to do the correct backlight calculation. So I think you were smart (or the Unity devs were smart?) to use a 64-sized pixbuf.

**Comment by SavageTiger** [comment]
@tliron Yeah after hours of digging through the documentation and C source I kinda came to the same conclusion as you. I just hoped that there was something I overlooked. There is only one case on my test machines where I see this happen, and thats with apps created using nativefier. I am not sure if its really that big of a deal.

I actually tried to read the pixels of the render stage, but that made Gnome hard-crash due to this bug: https://bugzilla.gnome.org/show_bug.cgi?id=653590

But like I said, I dont think its that big of a deal :)

**Comment by micheleg** [comment]
I'm kind of ready to merge (finally!) this feature. From my experiience it seemed to work very well. I've prepared a branch https://github.com/micheleg/dash-to-dock/tree/glossy-luanchers. Beside simple reorganizations and cleanups: 

1. I reimplemented the colorutils conversion functions. The reason is that, even though those code snipppets seem to be a bit everywhere on the internet, i could't find a clearly licensed source. I realized the math used is on Wikipedia so I just wuicly rewrote the functions. From my test the conversion is exactly the same. They are probably not as optimized as the originals, but we are not abusing of them. I hope I haven't introduced any bug, in partucular for some corner case.  I also moved the colorutils to a "namescape" rather than a class, as it doesn't make sense to instantiate an object all the time we want to use these utility functions.

2. I slightly changed the code computing the resampling factor, to be more generic. I aim at resampling down to 64 px as before. If for any reason the icon size is not specifically ~ 2^n (128, 256, etc), it will resample to something between 64 px and 128 px. I start resampling for size > 128 px, as - although I haven't tested it - I guess that the resampling plus dominance color calculations is not more efficient than calculating it directly (If someone wants to test this we can change thre details)

3. I slightly changed the updating functions, merging it with the dots one. Proper refactoring will hopefully follow.

I'm going to merge al the fixing/refactoring/performance commits and then proceed to merge into master if no problem or bugs is found.

**Comment by franglais125** [comment]
@micheleg Thanks for all the refactoring of the code!

At the moment, the feature is not really working for me. It seems that the color is just not there. Here are two screenshots:

Before refactoring:
screenshot from 2017-08-20 11-15-17

After refactoring:
screenshot from 2017-08-20 11-15-59


I'll try to see if I can find the issue. Did you find the same thing?

**Comment by micheleg** [comment]
Thanks a lot for the testing. It's working for me  (on 3.25.90). Maybe it depends on some specific settings.

**Comment by franglais125** [comment]
Ok, same as you... Work on 3.25.90 (Ubuntu), doesn't work on 3.22 (Debian 9). I'll keep investigating.

**Comment by franglais125** [comment]
Found the problem: I am using `isolate-monitors` on my Debian setup. I'll check the code for this.

**Comment by franglais125** [comment]
@micheleg Ok, following your fix it's all good on my side. I included a small fix here: https://github.com/franglais125/dash-to-dock/tree/glossy-luanchers, removing some trailing white spaces.

**Comment by micheleg** [comment]
Ok, added your fixes and reordered the commits. Thanks.

**Comment by SavageTiger** [comment]
Hurray, exciting! I have been running on my branch for weeks without any issues, I merged your latest changes, and as far as I can tell everything looks and feels the same on my machine. Been running it for a couple of hours now, did not detect any issues.

**Comment by micheleg** [comment]
I merged the feature in master [fe6dde50f0f5..5e2fd4c0c].

Thanks again to everybody.

**Comment by SavageTiger** [comment]
@micheleg I am not sure how much memory the iconCache actually uses. However, if someone for example has 205 active items (although admittedly that nobody should have that many active items), it will always recalculate the colorspace of atleast 5 items when they switch workspace. Although since we added resampling that is not a heavy burden anymore. But its something to keep in mind.

**Comment by micheleg** [comment]
Yes, given the limited memory footprint we can put the limit higher (for me 200 was a large number already). I just want to be sure it will never happen to grow to an uncontrolled size. My concern originates from the possibility that  some applications might get assigned inconsistent ids (e.g., I know that sometimes matlab figures do not get properly grouped under the same icon, which i guess implies different app id are used).   I'll put a higher limit then, with the expectation that we will never reach it.

**Comment by franglais125** [comment]
Perfect, I thought of the same thing with regards to "random" ids, but then couldn't come up with an example... Once you merge this patch (or a version of it), I'll try it over some long sessions to track its size (with some `global.log()` calls), and report back.

**Comment by micheleg** [comment]
Ok, merged in master as [b3f86f74be90a]. I put the limit to 1000.

**Comment by tliron** [comment]
@micheleg Thanks for rewriting the HSV/RGB functions! I adapted them for my project and credited you.

**Comment by franglais125** [comment]
@micheleg, reporting back: I know 5 days is not exactly long, but from my logs I haven't seen any "weird" app-id. I tried on purpose to open more apps and perform more actions than I would normally do, and I have the cache size at 25. Having the cache size limit at < 1000 seems to be more than ok.

**Comment by micheleg** [comment]
Good. Let's keep the check in any case, it just costs an if.
